### ISSUE TITLE
Using -S all syscall rule in combination with -k argument and libaudit.h functions

### ISSUE THREAD

**Issue by vhorvat** [linux-audit/audit-userspace#410]
I'm using a custom kernel (5.10), audit version 3.0.1, and I'm limited to this environment.

While trying to apply the -k (key) argument in a self-parsed rule using the audit_rule_fieldpair_data() function from the libaudit and the corresponding bitwise AUDIT_FILTER_MASK masking for the flags field, in a rule that doesn't have a specific syscall specified (using the "all" argument), I receive the EAU_KEYDEP error code.

Tracking the execution of the code, I see that the "all" argument is processed without setting the _audit_syscalladded variable to 1 in it's implementation in libaudit , which causes the failure when trying to set the key for that rule.

Example of the rule I'm trying to apply:

```
-a exit,always -F arch=b64 -S all -k some_key
-a exit,always -F arch=b32 -S all -k some_key
```

I can confirm that I am correctly parsing rules into the audit_rule_fieldpair_data() function, as it works with all rules that have syscalls explicitly specified. Mentioned rules also appear to work when -k opt is ommited. 

Is there other way of using the combination I'm trying to use (-S all with key) or is this indeed a limitation of the audit system?

I am also aware of the performance impact these rules may have on a system running them.
Thanks for support in advance!

**Comment by Cropi** [comment]
I am not sure if this is indeed an intended limitation, but taking a look at commit https://github.com/linux-audit/audit-userspace/commit/1482cec74f2d9472f81dd4f0533484bd0c26decd, I would say that `_audit_syscalladded = 1` could be added to `audit_rule_syscallbyname_data` to the beginning when there is a comparison with the "all" keyword.

**Comment by stevegrubb** [comment]
The libaudit rule API was designed exclusively for auditctl's own use because it ties closely with kernel work. Back in 2022, someone reported that their tool was having this same issue. Auditctl sets this itself. But with the realization that people are using this API, commit 1482cec was the result.

My recommendation is either set it in you code like auditctl is doing or add the patch mentioned above to a local build of the audit package. Either way, I think this can be closed since it was fixed in the 3.1 release.

**Comment by vhorvat** [comment]
Hi,

Thank you both for your comments.
As @stevegrubb provided additional clarity on the API usage and its purpose, along with a solution to my problem, I can now indeed close this issue.

Once again, thank you both for the fast feedback!

Best,
v.
